### PR TITLE
header: introduce ConstructFn type and integrate it into core.Listener and core.Exchange

### LINKED ISSUE #704: header: allow setting function that will create ExtendedHeader in header services
Currently our header creation is done by `MakeExtendedHeader` once new block is received(or block is requested by its hash). This restricts us from writing different tests(for example where  the header can be broken to verify the Fraud Service). 

As we discussed with @Wondertan , we can pass the exact function that will create our header and thus gives us flexibility to write different test cases.

### PR DESCRIPTION
**Pull request by vgonkivs**
Resolves #704

### CODE REVIEW

File: node/testing.go
Comment (Bidon15) [inline node/testing.go:38]: we also need to explicitly provide info on how the invalid blocks are created. Either it is from block N onwards. Or if it is the 1 of the M blocks produced.

File: node/testing.go
Comment (Bidon15) [inline node/testing.go:50]: in the dynamic set of fault headers, we need more explicit info log (can be included in the issue creation described below 👇 in the next comment review line)

**Review by Bidon15 — COMMENTED** [review]
Awesome idea with the fault maker 💪

File: node/testing.go
Comment (vgonkivs) [reply node/testing.go]: done

File: header/interface.go
Comment (vgonkivs) [reply header/interface.go:15]: added

File: node/testing.go
Comment (vgonkivs) [reply node/testing.go]: added

**Comment by Wondertan** [comment]
PR Naming suggestion/improvement: `header: introduce Maker type and integrate into core.Listener and core.Exchange`. There is also `FaultMaker` involved here and it seems now it could be part of a separate PR with its own PR, but it's fine.

File: ipld/test_helpers.go
Comment (Wondertan) [inline ipld/test_helpers.go:86]: * We must not ignore the error from the Commit and Import!!!
* If you make it as a testhelper, then pass the `t` and check for errors with t. Pls never ignore errors
* There is no need to pass EDS size as we have it in the `shares` already = `len(shares)^2`
* Use ipld.DefaultCodec.
* There is already a constructor for DAH in core/da pkg, which makes one from the EDS.
  * Lets better make this func similar to the `AddShares` as they are basically the same besides that one does rsmt2d.Compute and this one rsmt2d.Import and probably not make it a testhelper but a regular func that returns an error.  So we should call it ImportShares, place it below AddShares and make it return EDS.

File: node/testing.go
Comment (Wondertan) [inline node/testing.go:39]: `node.WithFaultMaker` name does not explain what exactly is getting faulted. Add Header to the option name

File: node/testing.go
Comment (Wondertan) [inline node/testing.go:74]: When I proposed an idea to make an option to set _custom_ Maker I meant setting a _custom_ Maker, not specifically a Faulty Maker.  To be honest, It feels strange to me to make an option specifically for one test case in the Swamp, even though it can be later enabled in the Testgound as well(cc @Bidon15).

Let's:
* Make an option `WithHeaderMaker(header.Maker)` which allows setting a custom header.Maker via fx.Replace. It should not be part of the testing.go in the node pkg, but a regular option.
* Extract the faultMaking logic into `header/testing.go` as `FraudMaker`, so that (1) you can import it in the Swamp test, @Bidon15 can import it in testgound tests and (2) its clear by the name that this maker causes FraudProof, not some arbitrary fault.

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: ipld/test_helpers.go
Comment (vgonkivs) [reply ipld/test_helpers.go]: done

File: ipld/add.go
Comment (renaynay) [inline ipld/add.go:41]: what does this comment mean?

File: ipld/add.go
Comment (renaynay) [inline ipld/add.go:56]:
```suggestion
	// commit the batch to DAG
```

**Review by renaynay — COMMENTED** [review]
minor nits but looks awesome!

File: header/core/exchange.go
Comment (vgonkivs) [reply header/core/exchange.go:22]: I'm ok with this.

File: node/testing.go
Comment (vgonkivs) [reply node/testing.go]: FaultMaker will be added in a different PR.

File: ipld/add.go
Comment (Wondertan) [inline ipld/add.go:41]:
```suggestion
		return nil, fmt.Errorf("ipld: importing empty data")
```

**Review by Wondertan — COMMENTED** [review]
LGTM.

@renaynay, I don't really get the Extender thing and we should discuss this. It is not clear what are we extending to call it Extender. There is one process of extending within our project and this is an extending of square. There is also an ExtendedHeader which means it has some additional value to the base Core header. To what Extending we are referring here specifically and if both at the same time, isn't this confusing?  

Previous Maker made sense to me because Maker is a synonym for the constructor. And it was constructing new Header. Now, it is not really clear.

File: ipld/add.go
Comment (vgonkivs) [reply ipld/add.go:41]: Alternatively, I can suggest `Converter` as it converts shares from` core.Block` to `ExtendedHeader`

Comment (Wondertan) [reply ipld/add.go:41]: Converting is about transforming something from one form into another. Here we convert multiple inputs into one, which is not really converting, and the resulting Header is not a cumulative of all the inputs, as it does not store has shares which are passed to the func in the form of a block. I would propose `header.Commiter` instead. It is not that broad as `Maker` and semantically close to blockchain/crypto terminology.

Comment (Wondertan) [reply ipld/add.go:41]: cc @renaynay

Comment (renaynay) [reply ipld/add.go:41]: I am in favour of either 

* `Generator`
* `Constructor` 

Either is fine with me bc encompasses the general thing that's happening within the function.

Comment (Wondertan) [reply ipld/add.go:41]: The function does much more stuff to be called `Constructor`, but `Generator` is fine. I would still be more precise with `Commiter` as it also involves what `Generator` means in some way and also makes it more connected to what we are doing. Basically, Bridge Node that executes this function commits to the correctness of the Header and distributes it on the network so others can verify the commitment/header. 

cc @liamsi, WDYT?

Comment (vgonkivs) [reply ipld/add.go:41]: `Commiter` and `Generator` are both ok for me

Comment (renaynay) [reply ipld/add.go:41]: `Generator`, lets just ship. I am not in favour of `Committer`. If we both like `Generator`, lets just compromise and use.

Comment (vgonkivs) [reply ipld/add.go]: Changed to generator

**Review by renaynay — APPROVED** [review]
👍🏻

File: ipld/add.go
Comment (liamsi) [reply ipld/add.go]: I'd even call it sth like `header.ConstructExtendedFn` or `header.CreateExtendedFn`(even the comment above the type definition says "a function that creates an ExtendedHeader" so the latte would call the type accordingly). You can skip the `Fn` suffix.


Just regarding the nuances of the verbs used here: 
generate/create is to bring something into being
construct is to build or form (something) by assembling parts

Comment (liamsi) [reply ipld/add.go]: Hlib mentioned to me that this is planned to become an interface. then the name should follow golang naming conventions for interfaces: 
> By convention, one-method interfaces are named by the method name plus an -er

IMO, we can do this once this will be an interface (and not necessarily in this PR).

Comment (renaynay) [reply ipld/add.go]: I personally like `Fn` suffix. It's descriptive and good 👍🏻 

I'm in favour of `header.ConstructExtendedFn`.

Comment (renaynay) [reply ipld/add.go]: `WithHeaderConstructorFn` but let's get @Wondertan approval first

Comment (Wondertan) [reply ipld/add.go]: Ok, let's do `header.ConstructFn`

Comment (vgonkivs) [reply ipld/add.go]: renamed

**Review by renaynay — APPROVED** [review]

File: node/settings.go
Comment (Wondertan) [inline node/settings.go:76]:
```suggestion
// WithHeaderConstructFn sets custom func that creates extended header
func WithHeaderConstructFn(construct header.ConstructFn) Option {
```

**Review by Wondertan — APPROVED** [review]
Apply the naming nit suggestion above, and we are ready to go

**Review by Bidon15 — APPROVED** [review]
Agree with Hlib's comment above 👆

**Pull request merged**
